### PR TITLE
Penn alma

### PR DESCRIPTION
**Pull request by jimfhahn**
Getting the Alma DAG process started. Alma API details have been added here: https://github.com/LD4P/ils-middleware/wiki/Alma-APIs-in-Airflow

### CODE REVIEW

File: ils_middleware/dags/penn.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+from ils_middleware.tasks.sinopia.local_metadata import new_local_admin_metadata
+from ils_middleware.tasks.sinopia.email import (
+    notify_and_log,
+    send_update_success_emails,
+)
+from ils_middleware.tasks.sinopia.login import sinopia_login
+from ils_middleware.tasks.sinopia.metadata_check import existing_metadata_check
+from ils_middleware.tasks.sinopia.rdf2marc import Rdf2Marc
+from ils_middleware.tasks.alma.new import NewMARCtoAlma
+from ils_middleware.tasks.alma.mod_xml import to_alma_xml
```
Comment (jermnelson) [inline ils_middleware/dags/penn.py:20]: Missing these task modules, could you add them to this PR?

File: ils_middleware/dags/penn.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    "on_failure_callback": task_failure_callback,
+}
+
+with DAG(
+    "penn",
+    default_args=default_args,
+    description="Penn Alma DAG",
+    schedule_interval=timedelta(minutes=5),
+    start_date=datetime(2021, 8, 24),
+    tags="alma",
```
Comment (jermnelson) [inline ils_middleware/dags/penn.py:51]: The DAG object is expecting a list of tags, to correct just change to `tags=["alma"],`

**Review by jermnelson — CHANGES_REQUESTED** [review]
Thanks @jimfhahn, this PR is missing the Alma task modules along with accompanying tests of those tasks that should be included in this PR. Also, the PR build is failing the lint, test, and type-checking steps. The README has directions for insuring that your PR will pass when running locally:

- [Testing]( https://github.com/LD4P/ils-middleware#automated-tests)
- [Linting](https://github.com/LD4P/ils-middleware#linting), running the **black** [code formatter](https://github.com/LD4P/ils-middleware#code-formatting) before running `flake8` usually corrects any linting errors 
- [Type Checking](https://github.com/LD4P/ils-middleware#typechecking)

Please let me know if you have any questions or run into any issues. Happy to pair to go over these steps if it would beneficially.

**Comment by jimfhahn** [comment]
@jermnelson feel free to take another look at this -- I managed to get the final test in a good place and pushed what I think is a working proof of concept for Penn Alma...

File: ils_middleware/tasks/amazon/alma_s3.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                    "bib"
+                )  # insert the <bib> root element required by alma
+                newroot.insert(0, root)
+                alma_xml = etree.tostring(
+                    newroot, xml_declaration=True, encoding="utf-8"
+                )
+                f = open("alma.xml", "wb")
+                f.write(alma_xml)
+                f.close()
+        s3_hook.load_bytes(
```
Comment (jermnelson) [inline ils_middleware/tasks/amazon/alma_s3.py:61]: Looking at the logic here, I think the `alma_xml` XML object only contains XML of the last record in the reader and not all of the MARC records in the marc_record.

File: ils_middleware/tasks/amazon/alma_s3.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                newroot = etree.Element(
+                    "bib"
+                )  # insert the <bib> root element required by alma
+                newroot.insert(0, root)
+                alma_xml = etree.tostring(
+                    newroot, xml_declaration=True, encoding="utf-8"
+                )
+                f = open("alma.xml", "wb")
+                f.write(alma_xml)
+                f.close()
```
Comment (jermnelson) [inline ils_middleware/tasks/amazon/alma_s3.py:60]: I'm not sure you need to write either XML streams to disk to do this XML manipulation. For example:

```python
from pymarc.marcxml import record_to_xml

marc_xml = record_to_xml(record)
tree = etree.fromstring(marc_xml)
newroot = etree.Element(
         "bib")  # insert the <bib> root element required by alma
newroot.insert(0, root)
```

**Review by jermnelson — CHANGES_REQUESTED** [review]

File: ils_middleware/tasks/alma/post.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        data=data,
+    )
+    logger.debug(f"alma result: {alma_result.status_code}\n{alma_result.text}")
+
+    # push the mms_id from the Alma API result to the next task, Sinopia Metadata Update
+    result = alma_result.content
+    fd = open("alma_api_response.xml", "wb")
+    fd.write(result)
+    fd.close()
+    xml_response = ET.parse("alma_api_response.xml")
```
[the remark this answers is not in the document]
Comment (jimfhahn) [reply ils_middleware/tasks/alma/post.py:66]: yes. that looks better, thanks!

File: alma.xml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+<?xml version='1.0' encoding='utf-8'?>
```
[the remark this answers is not in the document]
Comment (jimfhahn) [reply alma.xml]: yes, deleted the ones that were not needed, and moved the other to tests/fixtures directory

File: ils_middleware/tasks/amazon/alma_s3.py (hunk 4 above)
Comment (jimfhahn) [reply ils_middleware/tasks/amazon/alma_s3.py]: I like this, thanks!

File: ils_middleware/tasks/amazon/alma_s3.py (hunk 3 above)
Comment (jimfhahn) [reply ils_middleware/tasks/amazon/alma_s3.py]: I stopped using the XML write stream and replaced it with the pymarc method `record_to_xml`-- does that address the logic, or is there something else I should do here ?

**Review by jermnelson — APPROVED** [review]

**Pull request merged**